### PR TITLE
support network and machine light

### PR DESCRIPTION
**Pull request by rawdaGastan**
### Related Issues

- https://github.com/threefoldtech/tfgrid-sdk-go/issues/1213

### Try

- https://github.com/threefoldtech/tfgrid-sdk-go/blob/c59cf061a57a7b793c8e2d9034a30b5c8401fcb5/grid-client/scripts/deploy_single_vm_with_mycelium_light/main.go

### CODE REVIEW

File: grid-client/integration_tests/setup.go [hunk 1]
```diff
@@ -10,21 +10,20 @@ import (
...
 	"github.com/threefoldtech/tfgrid-sdk-go/grid-client/deployer"
 	"github.com/threefoldtech/tfgrid-sdk-go/grid-client/workloads"
-	"github.com/threefoldtech/zos/pkg/gridtypes"
+	"github.com/threefoldtech/tfgrid-sdk-go/grid-client/zos"
 	"golang.org/x/crypto/ssh"
 )
 
 func setup() (deployer.TFPluginClient, error) {
-	mnemonics := os.Getenv("MNEMONICS")
+	mnemonics := "winner giant reward damage expose pulse recipe manual brand volcano dry avoid"
```
Comment (ashraffouda) [inline grid-client/integration_tests/setup.go:23]: remove ur mnemonics

File: grid-client/integration_tests/setup.go [hunk 2]
```diff
@@ -10,21 +10,20 @@ import (
...
 	log.Printf("mnemonics: %s", mnemonics)
 
-	network := os.Getenv("NETWORK")
+	network := "dev"
```
Comment (ashraffouda) [inline grid-client/integration_tests/setup.go]: get the Network back from the env

**Review by ashraffouda — CHANGES_REQUESTED** [review]

File: grid-client/zos/deployment.go [hunk 3]
```diff
@@ -0,0 +1,308 @@
...
+			Requests:       requests,
+			WeightRequired: d.SignatureRequirement.WeightRequired,
+			Signatures:     signatures,
+			SignatureStyle: gridtypes.SignatureStyle(d.SignatureRequirement.SignatureStyle),
+		},
+		Workloads: workloads,
+	}
+}
+
+func (d *Deployment) zosDeployment4() *gridtypes4.Deployment {
```
Comment (xmonader) [inline grid-client/zos/deployment.go:254]: Should we make it `ZosDeployment4` to be consistent with `ZosDeployment`

File: grid-client/workloads/vm_light.go [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+
+	if sru == 0 {
+		return uint64(500 * gridtypes.Megabyte)
+	}
+
+	return uint64(2 * gridtypes.Gigabyte)
+}
+
+// LoadFromVM compares the vm with another given vm
+func (vm *VMLight) LoadFromVM(vm2 *VM) {
```
Comment (xmonader) [inline grid-client/workloads/vm_light.go]: should we support also a `LoadFromVMLight`?

**Review by xmonader — CHANGES_REQUESTED** [review]

File: grid-client/deployer/network_deployer.go [hunk 5]
```diff
@@ -36,387 +30,75 @@ func NewNetworkDeployer(tfPluginClient *TFPluginClient) NetworkDeployer {
...
+	if err != nil {
+		multiErr = multierror.Append(multiErr, err)
 	}
 
-	for _, znet := range znets {
-		dls, err := d.generateDeployments(znet, endpoints, nodeUsedPorts, publicNode)
+	for _, znet := range zNets {
+		deployment, err := znet.GenerateVersionlessDeployments(
+			ctx, d.tfPluginClient.NcPool, d.tfPluginClient.SubstrateConn,
+			d.tfPluginClient.TwinID, publicNode, allNodes, endpoints, nodeUsedPorts,
```
Comment (Eslam-Nawara) [inline grid-client/deployer/network_deployer.go:75]: why not passing the plugin instead

File: grid-client/deployer/network_deployer.go [hunk 6]
```diff
@@ -578,109 +278,92 @@ func (d *NetworkDeployer) updateStateFromDeployments(ctx context.Context, znet *
...
-			delete(znet.NodesIPRange, node)
+
+	var publicNode uint32
+	var err error
+	for nodeID := range allNodes {
+		// we check first that there is a public node in all nodes in the networks.
+		// that way we can save extra requests to get a new public node
+		// and its used ports and endpoints. this public node also might not be used
+		// as access node if there is no need for it.
+		if endpoints[nodeID] != nil && endpoints[nodeID].To4() != nil {
```
Comment (Eslam-Nawara) [inline grid-client/deployer/network_deployer.go:328]: if the network already has a public node here, we don't need to loop over all nodes to check if there's a public node right?

File: grid-client/state/state.go [hunk 7]
```diff
@@ -389,6 +400,76 @@ func (st *State) LoadNetworkFromGrid(ctx context.Context, name string) (znet wor
...
+	var nodes []uint32
+	nodesIPRange := map[uint32]zosTypes.IPNet{}
+	for _, net := range zNets {
+		maps.Copy(nodesIPRange, net.NodesIPRange)
+		nodes = append(nodes, net.Nodes...)
+	}
+
+	znet.NodeDeploymentID = nodeDeploymentsIDs
+	znet.Nodes = nodes
+	znet.NodesIPRange = nodesIPRange
```
Comment (Eslam-Nawara) [inline grid-client/state/state.go:472]: should we get mycelium keys too?

File: grid-client/zos/ipnet.go [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+
+type IPNet struct{ net.IPNet }
+
+// ParseIPNet parse iprange
+func ParseIPNet(txt string) (r IPNet, err error) {
+	if len(txt) == 0 {
+		//empty ip net value
+		return r, nil
+	}
+	//fmt.Println("parsing: ", string(text))
```
Comment (Eslam-Nawara) [inline grid-client/zos/ipnet.go]: remove this please

File: grid-client/zos/ipnet.go [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+// MarshalJSON dumps iprange as a string
+func (i IPNet) MarshalJSON() ([]byte, error) {
+	if len(i.IPNet.IP) == 0 {
+		return []byte(`""`), nil
+	}
+	v := fmt.Sprint("\"", i.String(), "\"")
+	return []byte(v), nil
+}
+
+// MarshalJSON dumps iprange as a string
```
Comment (Eslam-Nawara) [inline grid-client/zos/ipnet.go]: `MarshalText`

File: grid-client/zos/machine_light.go [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+	// - Exist, obviously
+	// - Not used by other VMs
+	// - Only possible on `dedicated` nodes
+	GPU []GPU `json:"gpu,omitempty"`
+}
+
+// MachineNetworkLight structure
+type MachineNetworkLight struct {
+	// Mycelium IP config, if planetary is true, but Mycelium is not set we fall back
+	// to yggdrasil support. Otherwise (if mycelium is set) a mycelium ip is used instead.
```
Comment (Eslam-Nawara) [inline grid-client/zos/machine_light.go:43]: no ygg here no?

File: grid-client/zos/workload.go [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+
+func (s ResultState) IsAny(state ...ResultState) bool {
+	for _, in := range state {
+		if s == in {
+			return true
+		}
+	}
+
+	return false
+}
```
Comment (Eslam-Nawara) [inline grid-client/zos/workload.go]:
```suggestion
	return slices.Contains(state, s)
}
```

File: grid-client/workloads/network_interface.go [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+	GenerateVersionlessDeployments(
+		ctx context.Context,
+		ncPool client.NodeClientGetter,
+		subConn subi.SubstrateExt,
+		twinID, publicNode uint32,
+		allNodes map[uint32]struct{},
+		endpoints map[uint32]net.IP,
+		nodeUsedPorts map[uint32][]uint16,
+	) (map[uint32]zos.Deployment, error)
+	generateDeployments(endpointIPs map[uint32]net.IP, usedPorts map[uint32][]uint16, publicNode, twinID uint32,
```
Comment (Eslam-Nawara) [inline grid-client/workloads/network_interface.go]: why is this added to the interface? isn't it just a helper function?

File: grid-client/zos/machine_light.go (hunk 10 above)
Comment (rawdaGastan) [reply grid-client/zos/machine_light.go:43]: no ygg or public ips

**Review by xmonader — APPROVED** [review]

**Review by ashraffouda — APPROVED** [review]

File: grid-client/state/state.go [hunk 13]
```diff
@@ -380,9 +380,15 @@ func (st *State) LoadNetworkFromGrid(ctx context.Context, name string) (znet wor
 		nodes = append(nodes, net.Nodes...)
 	}
 
+	myceliumKeys := make(map[uint32][]byte)
+	for _, net := range zNets {
+		maps.Copy(myceliumKeys, net.MyceliumKeys)
+	}
```
Comment (Eslam-Nawara) [inline grid-client/state/state.go]: momken n add this to the previous loop? you already loop over the zNets there

File: grid-client/state/state.go [hunk 14]
```diff
@@ -456,15 +462,21 @@ func (st *State) LoadNetworkLightFromGrid(ctx context.Context, name string) (zne
...
 	// merge networks
 	var nodes []uint32
-	nodesIPRange := map[uint32]zosTypes.IPNet{}
+	nodesIPRange := make(map[uint32]zosTypes.IPNet)
 	for _, net := range zNets {
 		maps.Copy(nodesIPRange, net.NodesIPRange)
 		nodes = append(nodes, net.Nodes...)
 	}
 
+	myceliumKeys := make(map[uint32][]byte)
```
Comment (Eslam-Nawara) [inline grid-client/state/state.go]: same here

**Review by Eslam-Nawara — APPROVED** [review]

**Pull request merged**